File: TP3DPBO2425C3/CPP/KodeProgram/Pegawai.cpp
// Pegawai.cpp
#include <string>
#include <vector>
#include <iostream>
#include <iomanip>
#include <stdexcept>

// Include dependensi dasar dan Interfaces
#include "RiwayatGaji.cpp" 
#include "IKepemimpinan.cpp"  
#include "IKehadiran.cpp" 

#ifndef PEGAWAI_CPP 
#define PEGAWAI_CPP

class Pegawai {
protected:
    std::string idPegawai;
    std::string nama;
    double gajiPokok;
    std::string alamatLengkap;
    std::string tanggalLahirString;

private:
    std::vector<RiwayatGaji> riwayatGaji; // Composition

public:
    // Konstruktor Dasar
    Pegawai(const std::string& id, const std::string& nama, double gaji, const std::string& alamat, const std::string& tglLahir);

    virtual ~Pegawai() = default; 

    // --- GETTER ---
    std::string getNama() const { return nama; }
    double getGajiPokok() const { return gajiPokok; }
    std::string getAlamatLengkap() const { return alamatLengkap; }
    std::string getTanggalLahirString() const { return tanggalLahirString; }
    const std::vector<RiwayatGaji>& getRiwayatGaji() const { return riwayatGaji; }

    // --- SETTER ---
    void setNama(const std::string& nama);
    void setGajiPokok(double gajiPokok);
    void setAlamatLengkap(const std::string& alamat);

    // Method Composition
    void tambahRiwayatGaji(const RiwayatGaji& riwayat);

    // Pure virtual functions (Polimorfisme)
    virtual void tampilkanData() const = 0;
    virtual double hitungTunjangan() const = 0;

protected:
    void tampilkanDasar() const;
};

// =================================================================
// IMPLEMENTASI METHOD DARI PEGAWAI
// =================================================================

// 1. Konstruktor
Pegawai::Pegawai(const std::string& id, const std::string& nama, double gaji, const std::string& alamat, const std::string& tglLahir)
    : idPegawai(id), nama(nama), gajiPokok(gaji), alamatLengkap(alamat), tanggalLahirString(tglLahir) {}

// 2. SETTER (Implementasi penuh Enkapsulasi)
void Pegawai::setNama(const std::string& nama) { 
    if (nama.empty()) throw std::invalid_argument("Nama tidak boleh kosong.");
    this->nama = nama; 
}
void Pegawai::setGajiPokok(double gajiPokok) { 
    if (gajiPokok < 3000000) throw std::invalid_argument("Gaji pokok minimal 3 juta.");
    this->gajiPokok = gajiPokok; 
}
void Pegawai::setAlamatLengkap(const std::string& alamat) { 
    this->alamatLengkap = alamat; 
}

// 3. Method Composition
void Pegawai::tambahRiwayatGaji(const RiwayatGaji& riwayat) {
    riwayatGaji.push_back(riwayat);
}

// 4. Method Tampil Dasar
void Pegawai::tampilkanDasar() const {
    std::cout << std::string(40, '=') << std::endl;
    std::cout << "ID Pegawai: " << idPegawai << " | Nama: " << nama << std::endl;
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "Gaji Pokok: " << gajiPokok << std::endl;
    std::cout << "Alamat Lengkap: " << alamatLengkap << std::endl;
    std::cout << "Tanggal Lahir: " << tanggalLahirString << std::endl;

    std::cout << "--- Riwayat Gaji (Composition) ---\n";
    for (const auto& riwayat : riwayatGaji) {
        riwayat.tampilkanData();
    }
    std::cout << std::string(40, '-') << std::endl;
}

#endif // PEGAWAI_CPP

File: TP3DPBO2425C3/Python/KodeProgram/Pegawai.py
# Pegawai.py
from abc import ABC, abstractmethod
from RiwayatGaji import RiwayatGaji

class Pegawai(ABC):
    """
    Kelas Abstrak/Parent: Pegawai. 
    Induk utama untuk Hierarchical Inheritance dan Composition.
    """
    def __init__(self, id_pegawai, nama, gaji_pokok, alamat, tgl_lahir):
        # Atribut Private
        self._id_pegawai = id_pegawai
        self._nama = nama
        self._gaji_pokok = gaji_pokok
        self._alamat_lengkap = alamat
        self._tanggal_lahir_string = tgl_lahir
        
        # Composition: Pegawai HAS-A RiwayatGaji[] (Array of Objects)
        self._riwayat_gaji = [] 

    # --- GETTER ---
    @property
    def id_pegawai(self): return self._id_pegawai
    @property
    def nama(self): return self._nama
    @property
    def gaji_pokok(self): return self._gaji_pokok
    @property
    def alamat_lengkap(self): return self._alamat_lengkap
    @property
    def riwayat_gaji(self): return self._riwayat_gaji 
    @property
    def tanggal_lahir_string(self): return self._tanggal_lahir_string 

    # --- SETTER ---
    @nama.setter
    def nama(self, value):
        if not value: raise ValueError("Nama tidak boleh kosong.")
        self._nama = value
    
    @gaji_pokok.setter
    def gaji_pokok(self, value):
        if value < 3000000: raise ValueError("Gaji pokok minimal 3 juta.")
        self._gaji_pokok = value

    @alamat_lengkap.setter
    def alamat_lengkap(self, value):
        self._alamat_lengkap = value

    def tambah_riwayat_gaji(self, riwayat: RiwayatGaji):
        """Method untuk ADD DATA"""
        self._riwayat_gaji.append(riwayat)

    @abstractmethod
    def tampilkan_data(self):
        self._tampilkan_dasar() 

    @abstractmethod
    def hitung_tunjangan(self):
        pass

    def _tampilkan_dasar(self):
        # OUTPUT BERSIH: Memastikan semua atribut dasar tampil
        print("="*40)
        print(f"ID Pegawai: {self.id_pegawai} | Nama: {self.nama}")
        print(f"Gaji Pokok: {self.gaji_pokok:,.2f}")
        print(f"Alamat Lengkap: {self.alamat_lengkap}")
        print(f"Tanggal Lahir: {self.tanggal_lahir_string}")
        
        # Tampilkan Riwayat Gaji (Composition)
        print("--- Riwayat Gaji (Composition) ---")
        for riwayat in self.riwayat_gaji:
            riwayat.tampilkan_data()
        print("-" * 40)